File: sdks/typescript/src/encryption.ts
// Copyright 2026 ConveyorQ
//
// SPDX-License-Identifier: Apache-2.0

import { createCipheriv, createDecipheriv, randomBytes } from "node:crypto";

/**
 * Encryptor seals and opens opaque byte slices end to end: a client encrypts a
 * payload before enqueue and a worker decrypts it on dispatch, so the server
 * stores and relays ciphertext only and holds no keys. Implementations may be
 * async (e.g. a KMS-backed codec); the built-in {@link AESGCM} is local.
 *
 * `decrypt` must accept any ciphertext a prior `encrypt` produced, including
 * ciphertext sealed under a now-retired key the implementation still holds, so
 * key rotation never strands stored data.
 */
export interface Encryptor {
  /** Seal plaintext into ciphertext that {@link Encryptor.decrypt} can open. */
  encrypt(plaintext: Uint8Array): Uint8Array | Promise<Uint8Array>;
  /** Open ciphertext produced by {@link Encryptor.encrypt}. */
  decrypt(ciphertext: Uint8Array): Uint8Array | Promise<Uint8Array>;
}

/**
 * Key is a named AES-256 secret. The id labels ciphertext so the matching
 * secret can be found again after the active key rotates; it travels in the
 * clear and must not itself be sensitive.
 */
export interface Key {
  /** Non-empty, at most 255 bytes. */
  id: string;
  /** 32-byte AES-256 key material. */
  secret: Uint8Array;
}

/** Thrown when a key cannot be used to build an Encryptor. */
export class InvalidKeyError extends Error {
  constructor(message: string) {
    super(`encryption: invalid key: ${message}`);
    this.name = "InvalidKeyError";
  }
}

/** Thrown when decrypting ciphertext sealed under a key the Encryptor does not hold. */
export class UnknownKeyIdError extends Error {
  constructor(keyId: string) {
    super(`encryption: unknown key id ${JSON.stringify(keyId)}`);
    this.name = "UnknownKeyIdError";
  }
}

/** Thrown when ciphertext is truncated or does not match the framing. */
export class MalformedCiphertextError extends Error {
  constructor(message: string) {
    super(`encryption: malformed ciphertext: ${message}`);
    this.name = "MalformedCiphertextError";
  }
}

/** Thrown when ciphertext fails its authentication tag (tampered or wrong key). */
export class AuthenticationError extends Error {
  constructor() {
    super("encryption: authentication failed");
    this.name = "AuthenticationError";
  }
}

// AES-256-GCM framing parameters — byte-compatible with the Go SDK so a task
// sealed by one SDK opens in the other.
const SCHEME_VERSION = 1;
const KEY_BYTES = 32;
const NONCE_BYTES = 12;
const TAG_BYTES = 16;
const MAX_KEY_ID_BYTES = 255;
const HEADER_FIXED_BYTES = 2;

const idEncoder = new TextEncoder();
const idDecoder = new TextDecoder();

/**
 * AESGCM is the built-in {@link Encryptor}: AES-256-GCM with a fresh random
 * nonce per call and a key id framed into each ciphertext. It holds a keyring
 * so a retired key still opens data sealed under it, and seals new data under
 * the active key. Use {@link newAESGCM} to construct one.
 *
 * Ciphertext layout (matching the Go SDK): version byte, key-id length byte,
 * key id, 12-byte nonce, ciphertext, 16-byte GCM tag. The version+id header is
 * authenticated as additional data, binding the key id to the result.
 */
export class AESGCM implements Encryptor {
  private readonly ring: Map<string, Uint8Array>;
  private readonly activeId: string;
  private readonly activeSecret: Uint8Array;

  private constructor(ring: Map<string, Uint8Array>, activeId: string) {
    this.ring = ring;
    this.activeId = activeId;
    this.activeSecret = ring.get(activeId)!;
  }

  /** @internal use {@link newAESGCM}. */
  static build(activeKeyId: string, keys: Key[]): AESGCM {
    if (activeKeyId === "") {
      throw new InvalidKeyError("active key id is empty");
    }

    if (keys.length === 0) {
      throw new InvalidKeyError("no keys provided");
    }

    const ring = new Map<string, Uint8Array>();

    for (const key of keys) {
      if (key.id === "") {
        throw new InvalidKeyError("empty key id");
      }

      if (idEncoder.encode(key.id).length > MAX_KEY_ID_BYTES) {
        throw new InvalidKeyError(`key id ${JSON.stringify(key.id)} exceeds ${MAX_KEY_ID_BYTES} bytes`);
      }

      if (key.secret.length !== KEY_BYTES) {
        throw new InvalidKeyError(`key ${JSON.stringify(key.id)} secret is ${key.secret.length} bytes, want ${KEY_BYTES}`);
      }

      if (ring.has(key.id)) {
        throw new InvalidKeyError(`duplicate key id ${JSON.stringify(key.id)}`);
      }

      ring.set(key.id, key.secret);
    }

    if (!ring.has(activeKeyId)) {
      throw new InvalidKeyError(`active key id ${JSON.stringify(activeKeyId)} not among keys`);
    }

    return new AESGCM(ring, activeKeyId);
  }

  encrypt(plaintext: Uint8Array): Uint8Array {
    const header = encodeHeader(this.activeId);
    const nonce = randomBytes(NONCE_BYTES);

    const cipher = createCipheriv("aes-256-gcm", this.activeSecret, nonce);
    cipher.setAAD(header);
    const body = Buffer.concat([cipher.update(plaintext), cipher.final()]);
    const tag = cipher.getAuthTag();

    return Buffer.concat([header, nonce, body, tag]);
  }

  decrypt(ciphertext: Uint8Array): Uint8Array {
    const { keyId, headerLength } = decodeHeader(ciphertext);

    const secret = this.ring.get(keyId);
    if (secret === undefined) {
      throw new UnknownKeyIdError(keyId);
    }

    const rest = ciphertext.subarray(headerLength);
    if (rest.length < NONCE_BYTES + TAG_BYTES) {
      throw new MalformedCiphertextError("missing nonce or tag");
    }

    const nonce = rest.subarray(0, NONCE_BYTES);
    const body = rest.subarray(NONCE_BYTES, rest.length - TAG_BYTES);
    const tag = rest.subarray(rest.length - TAG_BYTES);
    const header = ciphertext.subarray(0, headerLength);

    const decipher = createDecipheriv("aes-256-gcm", secret, nonce);
    decipher.setAAD(header);
    decipher.setAuthTag(tag);

    try {
      return Buffer.concat([decipher.update(body), decipher.final()]);
    } catch {
      throw new AuthenticationError();
    }
  }
}

/**
 * newAESGCM builds an {@link AESGCM} that seals under the key identified by
 * `activeKeyId` and can open data sealed under any key in `keys`. Pass several
 * keys to support rotation: keep the old key to decrypt existing data while a
 * new active key seals fresh data.
 */
export function newAESGCM(activeKeyId: string, ...keys: Key[]): AESGCM {
  return AESGCM.build(activeKeyId, keys);
}

function encodeHeader(keyId: string): Uint8Array {
  const idBytes = idEncoder.encode(keyId);
  const header = new Uint8Array(HEADER_FIXED_BYTES + idBytes.length);
  header[0] = SCHEME_VERSION;
  header[1] = idBytes.length;
  header.set(idBytes, HEADER_FIXED_BYTES);

  return header;
}

function decodeHeader(ciphertext: Uint8Array): { keyId: string; headerLength: number } {
  if (ciphertext.length < HEADER_FIXED_BYTES) {
    throw new MalformedCiphertextError("shorter than header");
  }

  if (ciphertext[0] !== SCHEME_VERSION) {
    throw new MalformedCiphertextError(`unsupported version ${ciphertext[0]}`);
  }

  const idLength = ciphertext[1]!;
  const headerLength = HEADER_FIXED_BYTES + idLength;

  if (ciphertext.length < headerLength) {
    throw new MalformedCiphertextError("truncated key id");
  }

  const keyId = idDecoder.decode(ciphertext.subarray(HEADER_FIXED_BYTES, headerLength));

  return { keyId, headerLength };
}

/** Metadata key set on an encrypted payload so a worker knows to decrypt it. */
export const ENCRYPTION_MARKER_KEY = "conveyor.encryption";
/** Metadata marker value identifying the framing version. */
export const ENCRYPTION_MARKER_VALUE = "1";

File: sdks/python/src/conveyorq/encryption.py
# Copyright 2026 ConveyorQ
#
# SPDX-License-Identifier: Apache-2.0

"""End-to-end payload encryption, byte-compatible with the Go and TypeScript SDKs.

A client seals a payload before enqueue and a worker opens it on dispatch, so
the server stores and relays ciphertext only and holds no keys. The built-in
:class:`AESGCM` codec frames each ciphertext as: a version byte, a key-id length
byte, the key id, a 12-byte nonce, the ciphertext, and the 16-byte GCM tag. The
version + key-id header is authenticated as additional data, binding the key id
to the result. This layout matches the other SDKs, so a task sealed by one opens
in any other.
"""

from __future__ import annotations

import os
from typing import Dict, Protocol, runtime_checkable

from cryptography.exceptions import InvalidTag
from cryptography.hazmat.primitives.ciphers.aead import AESGCM as _AESGCM

# Framing parameters -- must stay identical across SDKs.
_SCHEME_VERSION = 1
_KEY_BYTES = 32
_NONCE_BYTES = 12
_TAG_BYTES = 16
_MAX_KEY_ID_BYTES = 255
_HEADER_FIXED_BYTES = 2

#: Metadata key set on an encrypted payload so a worker knows to decrypt it.
ENCRYPTION_MARKER_KEY = "conveyor.encryption"
#: Metadata marker value identifying the framing version.
ENCRYPTION_MARKER_VALUE = "1"


@runtime_checkable
class Encryptor(Protocol):
    """Seals and opens opaque byte strings end to end.

    A client encrypts a payload before enqueue and a worker decrypts it on
    dispatch. :meth:`decrypt` must accept any ciphertext a prior :meth:`encrypt`
    produced -- including ciphertext sealed under a now-retired key the
    implementation still holds -- so key rotation never strands stored data.
    """

    def encrypt(self, plaintext: bytes) -> bytes:
        """Seal plaintext into ciphertext that :meth:`decrypt` can open."""
        ...

    def decrypt(self, ciphertext: bytes) -> bytes:
        """Open ciphertext produced by :meth:`encrypt`."""
        ...


class Key:
    """A named AES-256 secret.

    The ``id`` labels ciphertext so the matching secret can be found after the
    active key rotates; it travels in the clear and must not itself be sensitive.
    The ``secret`` is 32 bytes of AES-256 key material.
    """

    __slots__ = ("id", "secret")

    def __init__(self, id: str, secret: bytes) -> None:
        self.id = id
        self.secret = secret


class InvalidKeyError(Exception):
    """Raised when a key cannot be used to build an :class:`AESGCM`."""

    def __init__(self, message: str) -> None:
        super().__init__(f"encryption: invalid key: {message}")


class UnknownKeyIdError(Exception):
    """Raised when decrypting ciphertext sealed under a key the codec does not hold."""

    def __init__(self, key_id: str) -> None:
        super().__init__(f"encryption: unknown key id {key_id!r}")
        self.key_id = key_id


class MalformedCiphertextError(Exception):
    """Raised when ciphertext is truncated or does not match the framing."""

    def __init__(self, message: str) -> None:
        super().__init__(f"encryption: malformed ciphertext: {message}")


class AuthenticationError(Exception):
    """Raised when ciphertext fails its authentication tag (tampered or wrong key)."""

    def __init__(self) -> None:
        super().__init__("encryption: authentication failed")


class AESGCM:
    """The built-in :class:`Encryptor`: AES-256-GCM with a per-call random nonce.

    It holds a keyring so a retired key still opens data sealed under it, and
    seals new data under the active key. Construct one with :func:`new_aes_gcm`.
    """

    __slots__ = ("_ring", "_active_id", "_active")

    def __init__(self, ring: Dict[str, bytes], active_id: str) -> None:
        self._ring = ring
        self._active_id = active_id
        self._active = _AESGCM(ring[active_id])

    def encrypt(self, plaintext: bytes) -> bytes:
        """Seal ``plaintext`` under the active key with a fresh random nonce."""
        header = _encode_header(self._active_id)
        nonce = os.urandom(_NONCE_BYTES)
        body = self._active.encrypt(nonce, plaintext, header)

        return header + nonce + body

    def decrypt(self, ciphertext: bytes) -> bytes:
        """Open ``ciphertext``, selecting the key named in its header."""
        key_id, header_length = _decode_header(ciphertext)

        secret = self._ring.get(key_id)
        if secret is None:
            raise UnknownKeyIdError(key_id)

        rest = ciphertext[header_length:]
        if len(rest) < _NONCE_BYTES + _TAG_BYTES:
            raise MalformedCiphertextError("missing nonce or tag")

        nonce = rest[:_NONCE_BYTES]
        body = rest[_NONCE_BYTES:]
        header = ciphertext[:header_length]

        try:
            return _AESGCM(secret).decrypt(nonce, body, header)
        except InvalidTag:
            raise AuthenticationError() from None


def new_aes_gcm(active_key_id: str, *keys: Key) -> AESGCM:
    """Build an :class:`AESGCM` sealing under ``active_key_id`` and opening any of ``keys``.

    Pass several keys to support rotation: keep an old key to decrypt existing
    data while a new active key seals fresh data.
    """
    if active_key_id == "":
        raise InvalidKeyError("active key id is empty")

    if not keys:
        raise InvalidKeyError("no keys provided")

    ring: Dict[str, bytes] = {}

    for key in keys:
        if key.id == "":
            raise InvalidKeyError("empty key id")

        if len(key.id.encode("utf-8")) > _MAX_KEY_ID_BYTES:
            raise InvalidKeyError(f"key id {key.id!r} exceeds {_MAX_KEY_ID_BYTES} bytes")

        if len(key.secret) != _KEY_BYTES:
            raise InvalidKeyError(f"key {key.id!r} secret is {len(key.secret)} bytes, want {_KEY_BYTES}")

        if key.id in ring:
            raise InvalidKeyError(f"duplicate key id {key.id!r}")

        ring[key.id] = key.secret

    if active_key_id not in ring:
        raise InvalidKeyError(f"active key id {active_key_id!r} not among keys")

    return AESGCM(ring, active_key_id)


def _encode_header(key_id: str) -> bytes:
    id_bytes = key_id.encode("utf-8")

    return bytes((_SCHEME_VERSION, len(id_bytes))) + id_bytes


def _decode_header(ciphertext: bytes) -> tuple[str, int]:
    if len(ciphertext) < _HEADER_FIXED_BYTES:
        raise MalformedCiphertextError("shorter than header")

    if ciphertext[0] != _SCHEME_VERSION:
        raise MalformedCiphertextError(f"unsupported version {ciphertext[0]}")

    id_length = ciphertext[1]
    header_length = _HEADER_FIXED_BYTES + id_length

    if len(ciphertext) < header_length:
        raise MalformedCiphertextError("truncated key id")

    key_id = ciphertext[_HEADER_FIXED_BYTES:header_length].decode("utf-8")

    return key_id, header_length
